filtergen.8
.\" -*- nroff -*-
.TH FILTERGEN 8 "January 20, 2005"

.SH NAME

filtergen \- packet filter compiler

.SH SYNOPSIS

\fBfiltergen\fR [ \fB-h\fR | \fB--help\fR ] [ \fB-V\fR | \fB--version\fR ] [ \fB-c\fR | \fB--compile\fR ] [ \fB-r\fR | \fB--no-resolve\fR ] [ \fB-s \fIsource\fB\fR | \fB--source=\fIsource\fB\fR ] [ \fB-t \fItarget\fB\fR | \fB--target=\fItarget\fB\fR ] [ \fB-o \fIoutfile\fB\fR | \fB--output=\fIoutfile\fB\fR ] \fIinfile\fR

\fBfiltergen\fR [ \fB-h\fR | \fB--help\fR ] [ \fB-V\fR | \fB--version\fR ] [ \fB-c\fR | \fB--compile\fR ] [ \fB-r\fR | \fB--no-resolve\fR ] [ \fB-s \fIsource\fB\fR | \fB--source=\fIsource\fB\fR ] [ \fB-t \fItarget\fB\fR | \fB--target=\fItarget\fB\fR ] [ \fB-o \fIoutfile\fB\fR | \fB--output=\fIoutfile\fB\fR ] [ \fB-F \fIpolicy\fB\fR | \fB--flush=\fIpolicy\fB\fR ]

.SH DESCRIPTION

\fBfiltergen\fR compiles a high-level filtering description language into a
variety of target formats.

.SH USAGE

\fBfiltergen\fR reads the ruleset from the \fIinfile\fR specified on the
command line (or standard input if \fIinfile\fR is "\-") and outputs to
standard output (or \fIoutfile\fR) via an optionally specified backend.

.PP
Both short and GNU-style long options are accepted:

.TP
\fB-c, --compile\fR

Only try to "compile" the input, and do not generate any output.  This may be
useful to check that an input file has no syntax errors in it before one
attempts to use the result on a live server.

.TP
\fB-R, --no-resolve\fR

Don't try to resolve hostnames, protocol names, port names, icmp type
names, or any other human-readable names.  This is useful for
debugging where you can't guarantee the resolver will correctly
mogrify each name.  One side-effect of this option is that names that
resolve to two or more numbers normally won't be expanded, which may
affect what you let through or block on your packet filter.  Use with
caution.

.TP
\fB-s \fIsource\fB, --source=\fIsource\fB\fR

If specified, \fBfiltergen\fR will use the \fIsource\fR language
parser to read the input file, otherwise the default of
\fBfiltergen\fR will be used.  Supported source languages are
\fBfiltergen\fR.

.TP
\fB-t \fItarget-filter\fB, --target=\fItarget-filter\fB\fR

If specified, \fItarget-filter\fR will be used to select an output filter type,
otherwise the default of \fBiptables\fR will be used.  Supported backends
are \fBiptables\fR, \fBip6tables\fR, \fBiptables-restore\fR, \fBip6tables-restore\fR,
\fBipchains\fR, \fBipfilter\fR, \fBcisco\fR (for Cisco IOS access-lists), and \fBfiltergen\fR (to
re-emit the input in \fBfiltergen\fR's native syntax).

.TP
\fB-F \fIpolicy\fB, --flush=\fIpolicy\fB\fR

Flush mode.  Generate a set of rules for clearing all rules from the packet
filter.  Useful for firewall scripts that need to `shutdown' the firewall.
You can supply a \fIpolicy\fR argument in place of the usual filename, to
specify whether the flushed filter should default to \fBaccept\fR,
\fBreject\fR, or \fBdrop\fR.  It defaults to \fBaccept\fR, equivalent to
having no filter loaded at all.  It is not necessary to specify an
\fIinfile\fR when using flush mode.

.TP
\fB-o \fIoutfile\fB, --output=\fIoutfile\fB\fR

Write output to \fIoutfile\fR instead of standard output.

.TP
\fB-h, --help\fR
Show command help.

.TP
\fB-V, --version\fR
Show program version.

.SH BUGS

Not all backends implement all features.

The packet filter is not optimised.

.SH SEE ALSO

\fBfgadm\fR(8), \fBfilter_syntax\fR(5), \fBfilter_backends\fR(7)

.SH AUTHOR

\fBfiltergen\fR was originally written by Matthew Kirkwood.

Jamie Wilkinson <jaq@spacepants.org> then took on maintenance of the
project, added some features, and rewrote a lot of the internals.


HISTORY
0.13:
	Changed build system to SCons.
	Reorganised code structure.
	Added filtergen syntax emitter.
	Added option to turn off the resolver (useful for debugging!)
	Fixed bug with DROP targets at the end of forwaring chains.
	Added iptables-save input parser.

0.12.8:
    Fix some easy to cause crashes.
    Remove some obsolete code that caused long compilation times.
    Prevents inclusion of the root filesystem path directly.
    Test that the example programs compile.

0.12.7:
	Supports iptables-restore output format.
	Supports IPv6 names.
	Supports ip6tables and ip6tables-restore output format.
	Better warnings when services don't resolve.
	Increases include file recursion depth to 512 from 16.
	Directory includes now have stable ordering of the files within.
	
	Runs tests in parallel so their output is captured, and generate more output.
	Adds more compiler warnings.
	Compiles clean under clang.

0.12.6:
	Support shell globbing in include directives.
	Allow '*' as a network interface name.

	Modernise build environment:
	Fix many compiler warnings.
	Finish migration to Git as upstream host.

0.12.5:
	Small fixes to test correctness.

0.12.4:
	Added directory include support.
	Fixed relative-path include error in fgadm.
	Improved tests for include file parsing.

0.12.3:
	Fix critical resolver bug.

0.12.2:
	New resolver code removes need for struct casting
	Altered test suite to be more robust
	minor cosmetic bugs

0.12.1:
	fixed some dumb bugs
	fixed some compile bugs on 64 bit platforms

0.12:
	converted build system to automake
	rewrote parser
	added test suite
	add "oneway" option
	fix iptables log message format (append a space)
	add "fgadm" command
	multi-homed host support

0.11:
	fixed non-working example in filter_syntax man page
	fix 64-bit warning in filter.c netmask calculation
	add "-F [policy]" flush option
	better feedback on parse errors

0.10:
	some cleanups, slight memory usage reduction
	add log messages: 'log text "foo etc"'
	fixed iptables and ipchains brace-grouping
	fix filter_unroll loop-making bug
	add RPM spec file from Wil Cooley <wcooley@nakedape.cc>
	rename package to "filtergen"
	add init scripts and makefile
	add samples for router, host and proxy firewall

0.9:
	fix install target to install section 7 manpages too
	do protocol lookups properly, support protocol numbers
	lookup/normalise service names
	force normalisation of netmasks
	add "-m" to force service and host names and to be
		turned into port numbers and IP addresses

0.8:
	cleanup for iptables and ipchains to save reallocs
	fix input vs output thing in forwarding support for iptables
		and ipchains

0.7:
	add "local" and "forward" syntax to state that packets are only
		for local or remote consumption (and "-h" option to
		default rules to the former)
	begin manpages
	add "-o" option to specify output file
	fix "make install"
	getopt()ify filtergen.c
	add "-l" option to minimise the number of rules for established
		connections, implemented for iptables backend
	add more stuff to the iptables skeleton
	warning fixups
	implement "local" and "forward" for iptables backend
	add "-r" option to mirror "-h"
	cleanup ipchains driver somewhat
	implement "local" for ipchains backend ("forward" is not possible)
	fix bug where accept would get lost from:
		input eth0 { ! dest bar drop; accept; };
	iptables and ipchains add "established packet" rules to head of chain
	add "proxy" as an alias for "redirect"

0.6:
	disable negation of conjunctions ("or"s, aka brace-grouping) for now
	add "no skeleton" option
	fix a couple of segfaults/aborts where input filter is invalid
	allow literal +, ", and _ signs in input
	add sub-groups -- where supported, large sub-sections can be moved
		out-of-line to improve performance

0.5: that's better
	much fixing of iptables driver.  Now shouldn't leak
	like the proverbial sieve

0.4: testing begins
	add transproxy and masquerading support
	fix syntax problems with iptables and ipchains output
	simple logging support (logs only initiating packets)
	rule checking -- ports only allowed with tcp or udp
	initial icmp support

0.3:
	add ipfilter backend
	fix yet another silly bug in main()
	add warning notice to cisco driver
	cisco driver supports port ranges now

0.2: Various improvements
	rewrote iptables driver
	added ipchains driver
	added beginnings of Cisco IOS ACL driver
		(see TODO for why this doesn't work yet)
	added TODO and HONESTY files

0.1: Initial release
	iptables driver
	a few bugs


filtergen.c
/* packet filter compiler
 *
 * Copyright (c) 2002 Matthew Kirkwood
 * Copyright (c) 2003,2004 Jamie Wilkinson
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <stdarg.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include "filter.h"
#include "input/input.h"

static FILE *outfile;

/** Print the program usage to stderr. */
void usage(char *prog) {
  fprintf(stderr, "Usage: %s [-chVR] [-t backend] [-o output] input\n", prog);
  fprintf(stderr, "       %s [-chVR] [-t backend] [-o output] -F policy\n\n",
          prog);
  fprintf(stderr, "Options:\n");

#ifdef HAVE_GETOPT_H
  fprintf(stderr, " --compile/-c              compile only, no generate\n");
#else
  fprintf(stderr, "\t-c\t\tcompile only, no generate\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(stderr,
          " --no-resolve/-R           don't resolve hostnames or portnames\n");
#else
  fprintf(stderr, "\t-R\t\tdon't resolve hostnames or portnames\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(stderr,
          " --source/-s source        source language (default: filtergen)\n");
#else
  fprintf(stderr, "\t-s source\tsource language (default: filtergen)\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(
      stderr,
      " --target/-t target        generate for target (default: iptables)\n");
#else
  fprintf(stderr, "\t-t target\tgenerate for target (default: iptables)\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(
      stderr,
      " --flush/-F policy         don't process input, generate flush rules\n");
#else
  fprintf(stderr, "\t-F policy\tdon't process input, generate flush rules\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(stderr, " --output/-o filename      write the generated packet "
                  "filter to filename\n");
#else
  fprintf(stderr,
          "\t-o filename\twrite the generated packet filter to filename\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(stderr, " --help/-h                 show this help\n");
#else
  fprintf(stderr, "\t-h\t\tshow this help\n");
#endif

#ifdef HAVE_GETOPT_H
  fprintf(stderr, " --version/-V              show program version\n");
#else
  fprintf(stderr, "\t-V\t\tshow program version\n");
#endif
}

/** Output a string, and appends a newline, to the target file. */
int oputs(const char *s) {
  int r = 0;
  if (s) {
    r = fputs(s, outfile);
    if (r > 0) {
      fputc('\n', outfile);
      r++;
    }
  }
  return r;
}

/** Write a string, printf style, to the target file. */
int oprintf(const char *fmt, ...) {
  va_list args;
  va_start(args, fmt);
  return vfprintf(outfile, fmt, args);
}

/** Function pointer table containing source parsers */
struct source_parser_s {
  const char *name;
  source_parser *parser;
} source_parsers[] = {{"filtergen", filtergen_source_parser},
                      {"iptables-save", ipts_source_parser},
                      {NULL, NULL}};

/** Function pointer table containing target filter types and the emitter
 * functions */
struct filtyp {
  const char *name;
  sa_family_t family;
  filtergen *compiler;
  filter_flush *flusher;
} filter_types[] = {
    {"iptables", AF_INET, fg_iptables, flush_iptables},
    {"ip6tables", AF_INET6, fg_ip6tables, flush_ip6tables},
    {"iptables-restore", AF_INET, fg_iptrestore, flush_iptrestore},
    {"ip6tables-restore", AF_INET6, fg_ip6trestore, flush_ip6trestore},
    {"ipchains", AF_INET, fg_ipchains, flush_ipchains},
    {"ipfilter", AF_INET, fg_ipfilter, NULL},
    {"cisco", AF_INET, fg_cisco, NULL},
    {"filtergen", 0, emit_filtergen, NULL},
    {NULL, 0, NULL, NULL},
};

#ifdef HAVE_GETOPT_H
static struct option long_options[] = {{"help", no_argument, 0, 'h'},
                                       {"compile", no_argument, 0, 'c'},
                                       {"target", required_argument, 0, 't'},
                                       {"output", required_argument, 0, 'o'},
                                       {"flush", required_argument, 0, 'F'},
                                       {"version", no_argument, 0, 'V'},
                                       {"no-resolve", no_argument, 0, 'R'},
                                       {"source", required_argument, 0, 's'},
                                       {0, 0, 0, 0}};
#define GETOPT(x, y, z) getopt_long(x, y, z, long_options, NULL)
#else
#define GETOPT(x, y, z) getopt(x, y, z)
#endif

/** Program entry point. */
int main(int argc, char **argv) {
  struct filter *f;
  int l;
  char *filename = NULL, *ftn = NULL, *ofn = NULL, *source_name = NULL;
  struct source_parser_s *sp;
  struct filtyp *ft = NULL;
  int flags = 0;
  char *progname;
  int arg;
  enum filtertype flushpol = T_ACCEPT;

  progname = argv[0];

  while ((arg = GETOPT(argc, argv, "hco:t:F:VRs:")) > 0) {
    switch (arg) {
    case ':':
      usage(progname);
      exit(1);
      break;
    case 'h':
      usage(progname);
      exit(0);
      break;
    case 'c':
      flags |= FF_NOSKEL;
      break;
    case 'o':
      ofn = strdup(optarg);
      break;
    case 'R':
      flags |= FF_NORESOLVE;
      break;
    case 's':
      source_name = strdup(optarg);
      break;
    case 't':
      ftn = strdup(optarg);
      break;
    case 'F':
      flags |= FF_FLUSH;
      if (!strcasecmp(optarg, "accept")) {
        flushpol = T_ACCEPT;
      } else if (!strcasecmp(optarg, "drop")) {
        flushpol = DROP;
      } else if (!strcasecmp(optarg, "reject")) {
        flushpol = T_REJECT;
      } else {
        fprintf(stderr, "%s: flush policy unrecofgnised: %s\n", progname,
                optarg);
        usage(progname);
        exit(1);
      }
      break;
    case 'V':
      printf("filtergen " VERSION "\n");
      exit(0);
      break;
    default:
      break;
    }
  }
  if (!(flags & FF_FLUSH)) {
    if (optind >= argc) {
      usage(progname);
      exit(1);
    } else
      filename = argv[optind++];
  }

  if (ofn) {
    /* XXX - open a different tempfile, and rename on success */
    outfile = fopen(ofn, "w");
    if (!outfile) {
      fprintf(stderr, "%s: can't open output file \"%s\"\n", progname, ofn);
      return 1;
    }
  } else
    outfile = stdout;

  /* work out which source parser to use */
  if (!source_name || !*source_name)
    source_name = strdup("filtergen");
  for (sp = source_parsers; sp->name; ++sp) {
    if (!strcmp(source_name, sp->name))
      break;
  }
  if (!sp->name) {
    fprintf(stderr, "%s: source parser unrecognised: %s", progname,
            source_name);
    usage(progname);
    exit(1);
  }

  /* work out which target emitter to use */
  if (!ftn || !*ftn)
    ftn = strdup("iptables");
  for (ft = filter_types; ft->name; ft++)
    if (!strcmp(ftn, ft->name))
      break;
  if (!ft->name) {
    fprintf(stderr, "%s: target filter unrecognised: %s\n", progname, ftn);
    usage(progname);
    exit(1);
  }

  if (!ftn || !*ftn)
    ftn = strdup("iptables");
  for (ft = filter_types; ft->name; ft++)
    if (!strcmp(ftn, ft->name))
      break;
  free(ftn);
  if (!ft->name) {
    fprintf(stderr, "%s: target filter unrecognised: %s\n", progname, ftn);
    usage(progname);
    exit(1);
  }

  /* What to do, then? */
  if (flags & FF_FLUSH) {
    /* Just flush it */
    l = ft->flusher(flushpol);
  } else {
    /* Compile from a file */
    if (filename && !strcmp(filename, "-"))
      filename = NULL;

    struct filtergen_opts o;
    memset(&o, 0, sizeof o);
    o.family = ft->family;

    f = sp->parser(filename, flags, &o);
    l = ft->compiler(f, flags);
  }

  if (ofn)
    fclose(outfile);

  if (l < 0) {
    fprintf(
        stderr,
        "an error occurred: most likely the filters defined make no sense\n");
    if (ofn)
      unlink(ofn);
    return 1;
  }
  if (ofn) {
    free(ofn);
  }
  fprintf(stderr, "generated %d rules\n", l);
  return 0;
}
